tests/e2e/tests/pilot/mirror_test.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pilot
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/kube/inject"
+	"istio.io/istio/tests/util"
+	"istio.io/pkg/log"
+)
+
+func TestRouteMirroring(t *testing.T) {
+	cases := []struct {
+		name       string
+		yamlFile   string
+		percentage float64
+		// It's hard to predict a precise value of mirroring, once Envoy uses random numbers to apply the mirror
+		// https://www.envoyproxy.io/docs/envoy/v1.11.1/api-v2/api/v2/route/route.proto.html#route-routeaction-requestmirrorpolicy
+		threshold float64
+	}{
+		{
+			name:       "mirror-percent-absent",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-absent.yaml",
+			percentage: 100.0,
+			threshold:  0.0,
+		},
+		{
+			name:       "mirror-50%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-50.yaml",
+			percentage: 50.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-10%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-10.yaml",
+			percentage: 10.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-80%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-80.yaml",
+			percentage: 80.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-0%",
+			yamlFile:   "testdata/networking/v1alpha3/mirror-percent-0.yaml",
+			percentage: 0.0,
+			threshold:  0.0,
+		},
+	}
+
+	cfg := &deployableConfig{
+		Namespace:  tc.Kube.Namespace,
+		kubeconfig: tc.Kube.KubeConfig,
+	}
+	if err := cfg.Setup(); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err := cfg.Teardown(); err != nil {
+			log.Errorf("Error tearing down: %v", err)
+		}
+	}()
+
+	podsC := getAppPods(t, "c")
+	podsB := getAppPods(t, "b")
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+
+			if err := util.KubeApply(tc.Kube.Namespace, c.yamlFile, tc.Kube.KubeConfig); err != nil {
+				t.Fatalf("Error applying mirror yaml file: %v", err)
+			}
+			time.Sleep(5 * time.Second)
```

Comment:
This one bothers me a little, but I don't know if there's another way to handle this at the moment.

Reply:
resolved